[mlir][debug] Allow multiple DIGlobalVariableExpression on globals.

abidh opened the pull request:
Currently, we allow only one  DIGlobalVariableExpressionAttr per global. It is especially evident in import where we pick the first from the list and ignore the rest. In contrast, LLVM allows multiple DIGlobalVariableExpression to be attached to the global. They are needed for correct working of things like DICommonBlock. This PR removes this restriction in mlir. Changes are mostly mechanical. One thing on which I went a bit back and forth was the representation inside GlobalOp. I would be happy to change if there are better ways to do this.

abidh committed:
[mlir][debug] Allow multiple DIGlobalVariableExpression on globals.

Currently, we allow only one  DIGlobalVariableExpressionAttr per global.
Also when we are importing, we just pick the first one. In contrast,
LLVM allows multiple DIGlobalVariableExpression to be attached to
the global. They are needed for correct working of things like
DICommonBlock. This PR removes this restriction in mlir. The changes
are mostly mechanical. The testcases have been adjusted accordingly.

gysit commented on mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:
```diff
@@ -2089,7 +2089,7 @@ void GlobalOp::build(OpBuilder &builder, OperationState &result, Type type,
                      bool dsoLocal, bool threadLocal, SymbolRefAttr comdat,
                      ArrayRef<NamedAttribute> attrs,
-                     DIGlobalVariableExpressionAttr dbgExpr) {
+                     ArrayRef<Attribute> dbgExprs) {
```
```suggestion
                     ArrayRef<DIGlobalVariableExpressionAttr> dbgExprs) {
```
Can this be typed (here and in the other places)?

abidh replied:
I tried this but it did not work. The `DIGlobalVariableExpressionArrayAttr` becomes an ArrayAttr and that takes an `ArrayRef<Attribute>` and compiler does not allow `ArrayRef<DIGlobalVariableExpressionAttr>` to be passed for that. I thought doing manual conversion was not worth the gain.

gysit replied:
Ok makes sense! I think DIGlobalVariableExpressionArrayAttr has a verifier so things should ultimately be checked.

gysit commented on mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:
```diff
@@ -1055,44 +1055,50 @@ LogicalResult ModuleTranslation::convertGlobals() {
+    if (op.getDbgExprs()) {
+      for (auto attr : *op.getDbgExprs()) {
+        if (auto exprAttr =
+                dyn_cast_if_present<DIGlobalVariableExpressionAttr>(attr)) {
```
Can the debug expression array ever have a nullptr entry or contain something else than a `DIGlobalVariableExpressionAttr`? I would presume not and a `cast<DIGlobalVariableExpressionAttr>(attr) `makes more range. Maybe you can also use `op.getDbgExprs().getAsRange<DIGlobalVariableExpressionAttr>()`?

abidh replied:
Thanks for the suggestion. I have replaced it with `op.getDbgExprs()->getAsRange<DIGlobalVariableExpressionAttr>()`.

gysit commented on mlir/test/Target/LLVMIR/llvmir-debug.mlir:
```diff
@@ -690,3 +690,27 @@ llvm.func @test() {
+
+// -----
+
+// Test multiple DIGlobalVariableExpression on a global
```
```suggestion
// Test multiple DIGlobalVariableExpression on a global.
```
ultra nit:

abidh replied:
Thanks for catching this. Fixed.

gysit commented on mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td:
```diff
@@ -1152,6 +1153,9 @@ def LLVM_AddressOfOp : LLVM_Op<"mlir.addressof",
 }
 
+def DIGlobalVariableExpressionArrayAttr :
+  TypedArrayAttrBase<LLVM_DIGlobalVariableExpressionAttr, "an array of variable expressions">;
```
I would consider moving this right below the definition of LLVM_DIGlobalVariableExpressionAttr? We already have such arrays for example for LLVM_AliasScopeArrayAttr.

abidh replied:
Done.

gysit reviewed (commented):
Thanks! I added some nit comments.

I think the approach with a typed array attribute makes sense. There is already prior art for the alias analysis metadata. 

What alternative did you have in mind?

abidh commented:
> Thanks! I added some nit comments.
> 
> I think the approach with a typed array attribute makes sense. There is already prior art for the alias analysis metadata.
> 
> What alternative did you have in mind?

As I mentioned in the comment above, I was thinking if I could make the argument to the build function more strongly typed. But using `ArrayRef<DIGlobalVariableExpressionAttr>` did not work out.

abidh committed:
Handle review comments.

abidh committed:
Move declaration of DIGlobalVariableExpressionArrayAttr in different file.

As suggested in the review comments.

gysit commented on mlir/lib/Target/LLVMIR/ModuleImport.cpp:
```diff
@@ -914,14 +914,16 @@ LogicalResult ModuleImport::convertGlobal(llvm::GlobalVariable *globalVar) {
-  if (!globalExpressions.empty())
-    globalExpressionAttr =
-        debugImporter->translateGlobalVariableExpression(globalExpressions[0]);
+  for (auto expr : globalExpressions) {
```
```suggestion
  for (Attribute expr : globalExpressions) {
```
nit:

gysit reviewed (approved):
Thanks LGTM modulo nit!

abidh committed:
Update mlir/lib/Target/LLVMIR/ModuleImport.cpp

Co-authored-by: Tobias Gysi <tobias.gysi@nextsilicon.com>

abidh committed:
Fixed the type of variable.

I applied the suggested change but it caused a build failure as type of the auto variable was llvm::DIGlobalVariableExpression* and not Attribute. Fixed in this commit.

Also removed a stale comment.
